### DISCUSSION TITLE
General steps for adding a new server

### DISCUSSION THREAD

**Discussion by MatthewL246** [MatthewL246/pretendo-docker#63] [General]
Here's the general steps I do to set up a new server. Keep in mind that this is usually not an easy task.

1. Add the server as a Git submodule to `repos/`
2. Add a configuration for the server to `compose.yml`
3. Get the server to build and run inside Docker: setting up environment variables and config files correctly, and potentially making changes to the source code if necessary. If the server isn't Dockerized yet, write the Dockerfile.  General environment variables like ports go in `environment/server-name.env` and are committed to Git, secrets like the database password go in `environment/server-name.local.env` and are ignored by Git.
4. Add an entry for the new server to `scripts/update-account-servers-database.sh` and `scripts/run-in-container/update-account-servers-database.js` so the account server knows the game server exists
5. Various steps to make the server actually work (the hard part), including changing environment variables, configuration, and potentially the source code.
6. Update `scripts/setup-environment.sh` to write `server-name.local.env` correctly during setup
7. Use `git diff` to create patches in `patches/server-name` if you needed edit the source
8. Rerun `setup.sh` to reset your environment and ensure the new server works. One trick is to edit the `name: pretendo-network` in `compose.yml` to set a new name, which gives you a fresh server environment. Also, ensure that CI passes when you push your commit.

**Comment by Kratos26000** [comment]
I tried I do the command scripts/setup-environment and I got the file server.local.env I didn't want to put the perfect name and the thing is that when I run that deletes it .local.env I not include

**Reply by MatthewL246** [reply]
Yes, running `setup-environment.sh` deletes the `.local.env` files and then recreates them. You need to add functionality to the script to generate your `.local.env` file correctly based on other variables like database passwords.

**Reply by Kratos26000** [reply]
and I do how to put it after if it's really private to avoid everyone otherwise we can do by discord

**Reply by MatthewL246** [reply]
In general, secrets should be randomly generated. There are already functions in the script for doing that.

**Reply by Kratos26000** [reply]
it's a script you have to run or it's a good setup environment you have to run

**Reply by Kratos26000** [reply]
> @MatthewL246 sorry again i try error => CANCELED [website internal] load metadata for docker.io/library/node: How do I do this to display as they say ?

**Reply by MatthewL246** [reply]
I'm sorry, but I don't have the time to guide you through every step of the process.

**Reply by Kratos26000** [reply]
I tried without bothering you because I wanted to leave it alone and I tried but I couldn't get to the .local.env level and the patches because I didn't do it for mk7.
